File: devchaoskit/utils/state.py
"""
DevChaosKit — Engine state persistence.

Manages a .devchaos/state.json file that acts as the IPC bridge between
a running daemon (start --detach) and the stop / status commands.

The state file is written atomically (write-to-temp → rename) so readers
never see a partial write.
"""

from __future__ import annotations

import json
import os
import signal
from datetime import datetime, timezone
from pathlib import Path
from typing import Any, Optional

from devchaoskit.engine.chaos_engine import InjectionEvent


# ---------------------------------------------------------------------------
# Paths
# ---------------------------------------------------------------------------

STATE_DIR_NAME = ".devchaos"
STATE_FILE_NAME = "state.json"
LOG_FILE_NAME = "engine.log"


# ---------------------------------------------------------------------------
# State manager
# ---------------------------------------------------------------------------


class StateManager:
    """
    Reads and writes engine state to a local .devchaos/state.json file.

    Parameters
    ----------
    cwd:
        Root directory for the .devchaos/ folder. Defaults to the process
        working directory so each project keeps its own state.
    """

    def __init__(self, cwd: Optional[Path] = None) -> None:
        root = cwd or Path.cwd()
        self._dir = root / STATE_DIR_NAME
        self._state_file = self._dir / STATE_FILE_NAME
        self._log_file = self._dir / LOG_FILE_NAME

    # ------------------------------------------------------------------
    # Paths (public, for CLI consumers)
    # ------------------------------------------------------------------

    @property
    def state_file(self) -> Path:
        return self._state_file

    @property
    def log_file(self) -> Path:
        return self._log_file

    @property
    def state_dir(self) -> Path:
        return self._dir

    # ------------------------------------------------------------------
    # Write
    # ------------------------------------------------------------------

    def ensure_dir(self) -> None:
        self._dir.mkdir(parents=True, exist_ok=True)

    def write(self, data: dict[str, Any]) -> None:
        """Atomically write *data* to the state file."""
        self.ensure_dir()
        tmp = self._state_file.with_suffix(".tmp")
        tmp.write_text(json.dumps(data, default=str, indent=2), encoding="utf-8")
        tmp.replace(self._state_file)

    def write_from_engine(
        self,
        pid: int,
        config_path: str,
        cycle_count: int,
        down_containers: set[str],
        history: list[InjectionEvent],
        dry_run: bool,
        cooldown_remaining: float = 0.0,
        cooldown_total: int = 0,
    ) -> None:
        """Convenience wrapper that serialises engine state into the state file."""
        last = history[-1] if history else None
        self.write(
            {
                "pid": pid,
                "started_at": datetime.now(tz=timezone.utc).isoformat(),
                "config_path": str(config_path),
                "dry_run": dry_run,
                "cycle_count": cycle_count,
                "down_containers": sorted(down_containers),
                "cooldown_remaining": round(cooldown_remaining, 1),
                "cooldown_total": cooldown_total,
                "last_event": _serialise_event(last),
                "history": [_serialise_event(e) for e in history[-50:]],
            }
        )

    # ------------------------------------------------------------------
    # Read
    # ------------------------------------------------------------------

    def read(self) -> Optional[dict[str, Any]]:
        """Return the parsed state dict, or None if no state file exists."""
        if not self._state_file.exists():
            return None
        try:
            return json.loads(self._state_file.read_text(encoding="utf-8"))
        except (json.JSONDecodeError, OSError):
            return None

    # ------------------------------------------------------------------
    # Process management
    # ------------------------------------------------------------------

    @staticmethod
    def is_alive(pid: int) -> bool:
        """Return True if a process with *pid* is running."""
        try:
            os.kill(pid, 0)  # Signal 0 = existence check only
            return True
        except ProcessLookupError:
            return False
        except PermissionError:
            return True  # Exists but owned by another user

    def running_pid(self) -> Optional[int]:
        """
        Return the PID from the state file if that process is still alive,
        otherwise return None (stale state).
        """
        state = self.read()
        if state is None:
            return None
        pid = state.get("pid")
        if pid and self.is_alive(int(pid)):
            return int(pid)
        return None

    # ------------------------------------------------------------------
    # Cleanup
    # ------------------------------------------------------------------

    def clear(self) -> None:
        """Remove the state file (called after a clean stop)."""
        self._state_file.unlink(missing_ok=True)


# ---------------------------------------------------------------------------
# Helpers
# ---------------------------------------------------------------------------


def _serialise_event(event: Optional[InjectionEvent]) -> Optional[dict]:
    if event is None:
        return None
    return {
        "timestamp": event.timestamp.isoformat(),
        "action": event.action,
        "target": event.target,
        "dry_run": event.dry_run,
        "result_status": event.result_status,
        "error": event.error,
    }

File: pkg/utils/state.go
package utils

import (
	"encoding/json"
	"os"
	"path/filepath"
	"syscall"
	"time"
)

const (
	StateDirName  = ".entropy"
	StateFileName = "state.json"
	LogFileName   = "engine.log"
)

type EventRecord struct {
	Timestamp    time.Time `json:"timestamp"`
	Action       string    `json:"action"`
	Target       string    `json:"target"`
	DryRun       bool      `json:"dry_run"`
	ResultStatus string    `json:"result_status"`
	Error        string    `json:"error"`
}

type EngineState struct {
	PID               int           `json:"pid"`
	StartedAt         time.Time     `json:"started_at"`
	ConfigPath        string        `json:"config_path"`
	DryRun            bool          `json:"dry_run"`
	CycleCount        int           `json:"cycle_count"`
	DownContainers    []string      `json:"down_containers"`
	CooldownRemaining float64       `json:"cooldown_remaining"`
	CooldownTotal     int           `json:"cooldown_total"`
	LastEvent         *EventRecord  `json:"last_event"`
	History           []EventRecord `json:"history"`
}

type StateManager struct {
	dir       string
	stateFile string
	logFile   string
}

func NewStateManager(cwd string) *StateManager {
	if cwd == "" {
		cwd = "."
	}
	dir := filepath.Join(cwd, StateDirName)
	return &StateManager{
		dir:       dir,
		stateFile: filepath.Join(dir, StateFileName),
		logFile:   filepath.Join(dir, LogFileName),
	}
}

func (s *StateManager) EnsureDir() error {
	return os.MkdirAll(s.dir, 0755)
}

func (s *StateManager) StateFile() string {
	return s.stateFile
}

func (s *StateManager) LogFile() string {
	return s.logFile
}

func (s *StateManager) Write(data *EngineState) error {
	if err := s.EnsureDir(); err != nil {
		return err
	}
	tmp := s.stateFile + ".tmp"
	b, err := json.MarshalIndent(data, "", "  ")
	if err != nil {
		return err
	}
	if err := os.WriteFile(tmp, b, 0600); err != nil {
		return err
	}
	return os.Rename(tmp, s.stateFile)
}

func (s *StateManager) Read() (*EngineState, error) {
	b, err := os.ReadFile(s.stateFile)
	if err != nil {
		if os.IsNotExist(err) {
			return nil, nil
		}
		return nil, err
	}
	var data EngineState
	if err := json.Unmarshal(b, &data); err != nil {
		return nil, err
	}
	return &data, nil
}

func (s *StateManager) Clear() error {
	err := os.Remove(s.stateFile)
	if err != nil && !os.IsNotExist(err) {
		return err
	}
	return nil
}

func (s *StateManager) IsAlive(pid int) bool {
	proc, err := os.FindProcess(pid)
	if err != nil {
		return false
	}
	err = proc.Signal(syscall.Signal(0))
	if err == nil {
		return true
	}
	if err.Error() == "os: process already finished" || err.Error() == "no such process" {
		return false
	}
	return true
}

func (s *StateManager) RunningPID() *int {
	state, err := s.Read()
	if err != nil || state == nil {
		return nil
	}
	if s.IsAlive(state.PID) {
		return &state.PID
	}
	return nil
}
